src/libraries/System.Text.Json/tests/Common/JsonTestHelper.cs

```diff
@@ -17,59 +18,99 @@ public static void AssertJsonEqual(string expected, string actual)
 
-                    foreach (string name in propertyNames)
+                    foreach (var property in expectedProperties.Except(actualProperties))
                     {
-                        AssertJsonEqual(expected.GetProperty(name), actual.GetProperty(name));
+                        AssertTrue(false, $"Property \"{property}\" missing from actual object.");
+                    }
+
+                    foreach (var property in actualProperties.Except(expectedProperties))
+                    {
+                        AssertTrue(false, $"Actual object defines additional property \"{property}\".");
+                    }
+
+                    foreach (string name in expectedProperties)
+                    {
+                        path.Push(name);
+                        AssertJsonEqual(expected.GetProperty(name), actual.GetProperty(name), path);
+                        path.Pop();
                     }
                     break;
                 case JsonValueKind.Array:
-                    JsonElement.ArrayEnumerator expectedEnumerator = actual.EnumerateArray();
-                    JsonElement.ArrayEnumerator actualEnumerator = expected.EnumerateArray();
+                    JsonElement.ArrayEnumerator expectedEnumerator = expected.EnumerateArray();
+                    JsonElement.ArrayEnumerator actualEnumerator = actual.EnumerateArray();
 
+                    int i = 0;
                     while (expectedEnumerator.MoveNext())
                     {
-                        Assert.True(actualEnumerator.MoveNext());
-                        AssertJsonEqual(expectedEnumerator.Current, actualEnumerator.Current);
+                        AssertTrue(actualEnumerator.MoveNext(), "Actual array contains fewer elements.");
+                        path.Push(i++);
+                        AssertJsonEqual(expectedEnumerator.Current, actualEnumerator.Current, path);
+                        path.Pop();
                     }
 
-                    Assert.False(actualEnumerator.MoveNext());
+                    AssertTrue(!actualEnumerator.MoveNext(), "Actual array contains additional elements.");
                     break;
                 case JsonValueKind.String:
-                    Assert.Equal(expected.GetString(), actual.GetString());
+                    AssertTrue(expected.GetString() == actual.GetString());
                     break;
                 case JsonValueKind.Number:
                 case JsonValueKind.True:
                 case JsonValueKind.False:
                 case JsonValueKind.Null:
-                    Assert.Equal(expected.GetRawText(), actual.GetRawText());
+                    AssertTrue(expected.GetRawText() == actual.GetRawText());
                     break;
                 default:
                     Debug.Fail($"Unexpected JsonValueKind: JsonValueKind.{valueKind}.");
                     break;
             }
+
+            void AssertTrue(bool condition, string? message = null)
```

Comment:
nit: for easier parsing of the code,

```suggestion
            void AssertTrue(bool passCondition, string? message = null)
```

or something similar. Also use named argument syntax in the method calls.